lib/MySQL_Monitor.cpp

```diff
@@ -3280,6 +3288,63 @@ VALGRIND_ENABLE_ERROR_REPORTING;
 	return ret;
 }
 
+/**
+* @brief Processes the discovered servers to eventually add them to 'runtime_mysql_servers'.
+* @details This method takes a vector of discovered servers, compares them against the existing servers, and adds the new servers to 'runtime_mysql_servers'.
+* @param originating_server_hostname A string which denotes the hostname of the originating server, from which the discovered servers were queried and found.
+* @param discovered_servers A vector of servers discovered when querying the cluster's topology.
+* @param reader_hostgroup Reader hostgroup to which we will add the discovered servers.
+*/
+void MySQL_Monitor::process_discovered_topology(const std::string& originating_server_hostname, vector<MYSQL_ROW> discovered_servers, int reader_hostgroup) {
+	char *error = NULL;
+	int cols = 0;
+	int affected_rows = 0;
+	SQLite3_result *runtime_mysql_servers = NULL;
+
+	char *query=(char *)"SELECT DISTINCT hostname FROM monitor_internal.mysql_servers ORDER BY hostname";
+	proxy_debug(PROXY_DEBUG_ADMIN, 4, "%s\n", query);
+	monitordb->execute_statement(query, &error, &cols, &affected_rows, &runtime_mysql_servers);
+
+	if (error) {
+		proxy_error("Error on %s : %s\n", query, error);
+	} else {
+		vector<tuple<string, int, int>> new_servers;
+		vector<string> saved_hostnames;
+		saved_hostnames.push_back(originating_server_hostname);
+
+		// Do an initial loop through the query results to save existing runtime server hostnames
+		for (std::vector<SQLite3_row *>::iterator it = runtime_mysql_servers->rows.begin(); it != runtime_mysql_servers->rows.end(); it++) {
+			SQLite3_row *r1 = *it;
+			string current_runtime_hostname = r1->fields[0];
+
+			saved_hostnames.push_back(current_runtime_hostname);
+		}
+
+		// Loop through discovered servers and process the ones we haven't saved yet
+		for (MYSQL_ROW s : discovered_servers) {
+			string current_discovered_hostname = s[2];
+			string current_discovered_port = s[3];
+
+			if (find(saved_hostnames.begin(), saved_hostnames.end(), current_discovered_hostname) == saved_hostnames.end()) {
+				tuple<string, int, int> new_server(current_discovered_hostname, parseLong(current_discovered_port.c_str()), reader_hostgroup);
```

Comment:
The use of `parseLong` here together with the exception model doesn't improve the handling of a malformed response. There are other places in which the response from the backend is assumed not to be malformed, for example, in the number of rows returned above. If handling a invalid value for that response is desired, is much better to handle the case here, parsing the value into a local, and in case of failure, place a `proxy_error` with the details of the invalid row received. For safety, would be best to either clear the `new_servers` or return.

Reply:
Removing the exception handling here as we can rely on the other places where backend is assumed not to be malformed.